src/diracx/db/sandbox_metadata/db.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import annotations
+
+import datetime
+
+import sqlalchemy
+
+from ..utils import BaseDB
+from .schema import Base as SandboxMetadataDBBase
+from .schema import sb_Owners, sb_SandBoxes
+
+
+class SandboxMetadataDB(BaseDB):
+    # This needs to be here for the BaseDB to create the engine
+    metadata = SandboxMetadataDBBase.metadata
+
+    async def _get_put_owner(self, owner: str, owner_group: str) -> int:
+        """adds a new owner/ownerGroup pairs, while returning their ID if already existing
+
+        Args:
+            owner (str): _description_
+            owner_group (str): _description_
+        """
+        stmt = sqlalchemy.select(sb_Owners.OwnerID).where(
+            sb_Owners.Owner == owner, sb_Owners.OwnerGroup == owner_group
+        )
+        result = await self.conn.execute(stmt)
+        res = result.fetchall()
+        if res:
+            return res[0][0]
+
+        stmt = sqlalchemy.insert(sb_Owners).values(Owner=owner, OwnerGroup=owner_group)
+        result = await self.conn.execute(stmt)
+        return result.lastrowid
+
+    async def insert(
+        self, owner: str, owner_group: str, sb_SE: str, se_PFN: str, size: int = 0
+    ) -> tuple[int, bool]:
+        """_insert a new sandbox
+        this is "equivalent" of DIRAC registerAndGetSandbox
+
+        Args:
+            owner (str): _description_
+            owner_group (str): _description_
+            sb_SE (str): _description_
+            sb_PFN (str): _description_
+            size (int, optional): _description_. Defaults to 0.
+        """
+        owner_id = await self._get_put_owner(owner, owner_group)
+        stmt = sqlalchemy.insert(sb_SandBoxes).values(
+            OwnerId=owner_id, SEName=sb_SE, SEPFN=se_PFN, Bytes=size
+        )
+        try:
+            result = await self.conn.execute(stmt)
+            return result.lastrowid, True
+        except sqlalchemy.exc.IntegrityError:
+            # it is a duplicate, try to retrieve SBiD
+            stmt = sqlalchemy.select(sb_SandBoxes.SBId).where(  # type: ignore
+                sb_SandBoxes.SEPFN == se_PFN,
+                sb_SandBoxes.SEName == sb_SE,
+                sb_SandBoxes.OwnerId == owner_id,
+            )
+            result = await self.conn.execute(stmt)
+            res = result.fetchall()
+            if res:
+                sb_ID = res[0][0]
+                stmt = (
+                    sqlalchemy.update(sb_SandBoxes)  # type: ignore
+                    .where(sb_SandBoxes.SBId == sb_ID)
+                    .values(LastAccessTime=datetime.datetime.utcnow())
+                )
+                await self.conn.execute(stmt)
+                return sb_ID, False
+            return 0, False
```

Comment:
It's not obvious to me how this line is supposed to be reachable?

If it isn't reachable, the return type of the function can be simplified from `tuple[int, bool]` to `int`.

Reply:
It seems reachable to me.

Reply:
How? Either the row already exists or it gets added (and we're within a single DB transaction so we can't have a race).